Semantic types pt 1 (special-type -> semantic-type rename)

dpsutton opened the pull request:
Step one of our semantic type system refactor. This just does the huge amount of no actual changes: everything should just be called semantic type instead of special type.

The point of breaking this up is to get this huge part in so we don't keep chasing conflicts and so when we get to the eventual semantic (heh) differences they aren't lost in a sea of formatting and name changes in 150+ files.

dpsutton committed:
Semantic types migration

dpsutton committed:
Fixup for basic querying

dpsutton committed:
Remove the relation type migration.

it makes the diff far too hard at the moment. need to be able to just
move everything from special_type -> semantic_type and then correct
the few that look at it to care about the effective type.

the PK/FK stuff gets really invasive and needs to be in its own much
smaller changeset

dpsutton committed:
Just rename special_type -> semantic_type for first change

dpsutton committed:
Special type -> semantic type

dpsutton committed:
special-type -> semantic-type

dpsutton committed:
SpecialType -> SemanticType

dpsutton committed:
special type -> semantic type

mostly in documentation, docstrings, etc

dpsutton committed:
Fix tests which relied on order of sets

database position was annotated by map-index'ing over the
set. changing `:special-type` to `:semantic-type` changed the order of
the seq produced from it.

dpsutton committed:
special -> semantic in schema_metadata

paulrosenzweig reviewed (approved):
I looked over the FE files and didn't spot any potential renaming issues.

Also, I love the thoroughness! Even css classes and comments were updated!

paulrosenzweig commented:
One other thought: It might make sense to merge `release-x.38.x` into `master` before merging this. There's been a fair amount of work on the release branch and it'd be better to resolve special/semantic conflicts on this PR rather than in the `release-x.38.x` -> `master` PR.

dpsutton commented:
> One other thought: It might make sense to merge release-x.38.x into master

you're 100% right. Hopefully we can get that and then I can watch all in flight PRs to see if they introduce new usages or anything.

Thanks for such a close eye noticing the css stuff. I know a 180 file diff is almost comically hopeless to do anything meaningful in.

As of now I believe the only use of "special type" in any capacity is in the event that we publish. And that can certainly wait for a later day.

tsmacdonald commented on src/metabase/api/field.clj:
```diff
@@ -52,7 +52,7 @@
     ;; Normal read perms = normal access.
     ;;
-    ;; There's also aspecial case where we allow you to fetch a Field even if you don't have full read permissions for
+    ;; There's also asemantic case where we allow you to fetch a Field even if you don't have full read permissions for
```
```suggestion
    ;; There's also a special case where we allow you to fetch a Field even if you don't have full read permissions for
```

tsmacdonald commented on src/metabase/db/data_migrations.clj:
```diff
@@ -237,16 +237,16 @@
 (defmigration ^{:author "camsaul", :added "0.29.0"} mark-category-fields-as-list
   (db/update-where! Field {:has_field_values nil
-                           :special_type     (mdb.u/isa :type/Category)
+                           :semantic_type    (mdb.u/isa :type/Category)
```
I don't know how the timing between data and DB migrations work—is it correct for these old data migrations to use the new column name?

dpsutton replied:
really good question. the docstring from the ns seems to say it could be both:

>   These migrations are all ran once when Metabase is first launched, except when transferring data from an existing
  H2 database.  When data is transferred from an H2 database, migrations will already have been run against that data;
  thus, all of these migrations need to be repeatable, e.g.:

I'm not sure how to handle that besides catching errors and trying again.

tsmacdonald replied:
@camsaul have you run into this before?

tsmacdonald commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -237,43 +237,43 @@
   created, `value` here might be a raw value or a map including information about the Field it references as well as a
   value.) For numbers, dates, and the like, this will parse the string appropriately; for `text` parameters, this will
-  additionally attempt handle special cases based on the base type of the Field, for example, parsing params for UUID
+  additionally attempt handle semantic cases based on the base type of the Field, for example, parsing params for UUID
```
```suggestion
  additionally attempt handle special cases based on the base type of the Field, for example, parsing params for UUID
```

tsmacdonald commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -295,7 +295,7 @@
                                      (parse-value-for-type param-type v))})
 
-   ;; Field Filters with "special" base types
+   ;; Field Filters with "semantic" base types
```
```suggestion
   ;; Field Filters with "special" base types
```

dpsutton committed:
Tim is awesome: Undo overeager special->semantic in docstrings

tsmacdonald commented on src/metabase/driver/common/parameters/values.clj:
```diff
@@ -237,43 +237,43 @@
-  to parse it as appropriate based on the base type and special type of the Field associated with it). These are
-  special cases for handling types that do not have an associated parameter type (such as `date` or `number`), such as
+  to parse it as appropriate based on the base type and semantic type of the Field associated with it). These are
+  semantic cases for handling types that do not have an associated parameter type (such as `date` or `number`), such as
```
```suggestion
  special cases for handling types that do not have an associated parameter type (such as `date` or `number`), such as
```

tsmacdonald reviewed (approved):
Few minor comments

```
$ ack -i 'special.?ty' src/ frontend/test/metabase frontend/src/
frontend/src/metabase/admin/datamodel/components/database/ColumnItem.jsx
155:      "Update Field Special-Type",
```
^looks good (the one hit is discussed [here](https://metaboat.slack.com/archives/C505ZNNH4/p1613062794092300))

```
$ ack -i 'semantic(?!\.|-|_|:|(\s?type))' src/ frontend/test/metabase frontend/src/ -A 1
src/metabase/api/table.clj
308:  "This method clears the semantic_type attribute for PK/FK fields of nested queries. Those fields having a semantic
309-  type confuses the frontend and it can really used in the same way"

src/metabase/related.clj
41:          Relevant parts are those that carry semantic meaning, and especially
42-          context-bearing forms."

src/metabase/sync/analyze/classifiers/category.clj
26:      ;; Don't let IDs become list Fields (they already can't become categories, because they already have a semantic
27-      ;; type). It just doesn't make sense to cache a sequence of numbers since they aren't inherently meaningful

src/metabase/sync/analyze/query_results.clj
48:     ;; If we already know the semantic type, becouse it is stored, don't classify again, but try to refine semantic
49-     ;; type set upstream for aggregation cols (which come back as :type/Number).

src/metabase/driver/common/parameters/values.clj
257:  are used to determine what 'semantic' type interpretation is required (e.g. for UUID fields)."
258-  [{{base-type :base_type, semantic-type :semantic_type} :field, {value :value} :value, :as field-filter} :- FieldFilter]

frontend/src/metabase/lib/schema_metadata.js
89:  for (const prop of ["base", "semantic"]) {
90-    const allowedTypes = typeDefinition[prop];

```
^also looks good (searching for false positives). (Edited after Dan fixed the ones I found the first time 'round)

dpsutton commented:
yeah thanks so much Tim. I think there are two blockers here:
- need release merged in to master and them i'm gonna update with those changes. I expect some annoying conflicts but nothing too terrible. Probably pretty easy to take master's changes and then run the rename on the smaller diff.
- need to resolve the issue you pointed out in `metabase.db.data-migrations` as that's not obvious to me. I wasn't aware we had clojure based migrations in addition to our sql based ones.

dpsutton committed:
Un-rename semantic_type in data_migrations

These migrations are run unless a migrations table marks them as
already having run. If they haven't run, then the db is older and the
column is special_type most likely. So we let them run as
`:special_type` and add some error handling to the migration runner
that is _opt in_.

```clojure
(defmigration ^{:author "camsaul", :added "0.20.0", :catch? true}
  migrate-field-types
  ...
  )

(try
  (@migration-var)
  (catch Exception e
    (if catch? ;; catch? from metadata
      (log/warn (format "Data migration %s failed: %s" migration-name (.getMessage e)))
      (throw e))))
```

dpsutton committed:
Merge branch 'master' into semantic-types

just accept theirs and will reapply the changes needed

dpsutton committed:
Fix merged master changes

just accepted their changes and fixed up rather than fix conflicts.
